packages/fetchai/connections/ledger/connection.py

```diff
@@ -133,9 +133,13 @@ async def receive(self, *args, **kwargs) -> Optional["Envelope"]:
             done_task = self.done_tasks.pop()
             return self._handle_done_task(done_task)
 
-        if len(self.receiving_tasks) == 0:
-            await self.event_new_receiving_task.wait()
+        while len(self.receiving_tasks) == 0:
             self.event_new_receiving_task.clear()
+            try:
+                await self.event_new_receiving_task.wait()
+            except Exception:
+                return None
+        self.event_new_receiving_task.clear()
```

Comment:
Check it the list is empty. If so, clear the flag (that could be inadvertently set?).
Then, wait until the flag has been set (by the `send()` method). If something bad happens (e.g. the receive task is cancelled), return None.
Then, clear the flag again, because we know the receiving task list is not empty.